environments-preview feature not exchanging DOCKER_HOST

**Describe the bug**

Enabling `environments` breaks the docker build process in CI/CD and complains that the `DOCKER_HOST` is in the wrong place.

----

We have a `docker_image` that depends on a `pex_binary`

<details>
  <summary>BUILD</summary>

```python
pex_binary(
    name = "bin",
    dependencies = [
        ":lib",
    ],
    execution_mode = "venv",
    include_tools = True,
    layout = "packed",
)

docker_image(
    name = "docker",
    dependencies = [
        ":docker_resources",
        ":bin",
    ],
    image_tags = ["{build_args.PANTS_DISTRIBUTION_VERSION}"],
)
```

</details>

Due to some wheel compatibility issues, we cannot build this `pex_binary` (and subsequent `docker_image`) on our Macs, it only works up in CI/CD (self-hosted GitHub Actions runners) where the machine is `linux_x86`.

To resolve this issue for ourselves locally, we enabled the `environments` feature

<details>
  <summary>BUILD</summary>

```python
local_environment(
  name="local_linux_x86",
  description="Localhost x86 Linux Environment",
  compatible_platforms=["linux_x86_64"],
  fallback_environment="docker_x86",
)

docker_environment(
  name="local_docker_x86",
  description="Dockerized x86 Linux Environment",
  platform="linux_x86_64",
  image="python:3.8.16",
)
```

</details>

And we attached the environment to the `pex_binary`

<details>
  <summary>BUILD</summary>

```python
pex_binary(
    name = "bin",
    dependencies = [
        ":lib",
    ],
    execution_mode = "venv",
    include_tools = True,
    layout = "packed",
    environment = "local_linux_x86",
)
```

</details>

The idea here is that the `pex_binary` needs to be built on a `linux_x86_64` platform and when that platform isn't available locally it should just use the `docker_environment`.

This solution is working for us locally, but when we run it up in CI/CD it no longer works. Here is the issue that we're seeing in CI/CD:

<details>
  <summary>Log Output</summary>

```log
15:54:05.51 [DEBUG] Starting: Scheduling: Building docker image ***.dkr.ecr.us-east-1.amazonaws.com/moz-batch-job:16.0.0-beta.1
15:54:05.54 [DEBUG] Starting: acquire_command_runner_slot
15:54:19.81 [DEBUG] Completed: setup_sandbox
15:54:19.86 [DEBUG] spawned local process as Some(1133) for Process { argv: ["/usr/bin/docker", "build", "--pull=False", "--tag", "***.dkr.ecr.us-east-1.amazonaws.com/moz-helper-services:latest", "--tag", "***.dkr.ecr.us-east-1.amazonaws.com/moz-helper-services:16.0.0-beta.1", "--build-arg", "PANTS_DISTRIBUTION_VERSION", "--file", "services/moz-helper-services/Dockerfile", "."], env: {"PANTS_DISTRIBUTION_VERSION": "16.0.0-beta.1", "PATH": "/tmp/pants-sandbox-q6Zu6z/.shims/bin", "__UPSTREAM_IMAGE_IDS": ""}, working_directory: None, input_digests: InputDigests { complete: DirectoryDigest { digest: Digest { hash: Fingerprint<2c4455e78a6e491e19a07eca09e00a2fc9d026b6e428026680cb253cbbf4d541>, size_bytes: 264 }, tree: "Some(..)" }, nailgun: DirectoryDigest { digest: Digest { hash: Fingerprint<e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855>, size_bytes: 0 }, tree: "Some(..)" }, input_files: DirectoryDigest { digest: Digest { hash: Fingerprint<79b1e31b408edaf0ecb646a2deaf492f0d62436d46c7e020659821ef142b3191>, size_bytes: 184 }, tree: "Some(..)" }, immutable_inputs: {RelativePath(".shims"): DirectoryDigest { digest: Digest { hash: Fingerprint<e6c1927a092ff6beaa1fa4f372248c975cb51869b1c28245cb0281ebc6816ecb>, size_bytes: 78 }, tree: "Some(..)" }}, use_nailgun: {} }, output_files: {}, output_directories: {}, timeout: None, execution_slot_variable: None, concurrency_available: 0, description: "Building docker image ***.dkr.ecr.us-east-1.amazonaws.com/moz-helper-services:latest +1 additional tag.", level: Info, append_only_caches: {}, jdk_home: None, platform: Linux_x86_64, cache_scope: PerSession, execution_strategy: Local, remote_cache_speculation_delay: 0ns }
15:54:19.89 [INFO] Completed: Building docker image ***.dkr.ecr.us-east-1.amazonaws.com/moz-helper-services:latest +1 additional tag.
15:54:19.89 [DEBUG] Completed: Scheduling: Building docker image ***.dkr.ecr.us-east-1.amazonaws.com/moz-helper-services:latest +1 additional tag.
15:54:19.89 [DEBUG] Completed: acquire_command_runner_slot
15:54:19.89 [DEBUG] Running Building docker image ***.dkr.ecr.us-east-1.amazonaws.com/moz-result-processor:16.0.0-beta.1 under semaphore with concurrency id: 1, and concurrency: 1
15:54:19.89 [INFO] Starting: Building docker image ***.dkr.ecr.us-east-1.amazonaws.com/moz-result-processor:16.0.0-beta.1
15:54:19.89 [DEBUG] Starting: setup_sandbox
15:54:19.89 [WARN] Docker build failed for `docker_image` services/moz-helper-services:docker. The services/moz-helper-services/Dockerfile has `COPY` instructions for source files that may not have been found in the Docker build context.

However there are possible matches. Please review the following list of suggested renames:

  * services.moz-helper-services/bin.pex => services/moz-helper-services


15:54:19.90 [DEBUG] Completed: `publish` goal
15:54:19.90 [DEBUG] computed 1 nodes in 44.638043 seconds. there are 13299 total nodes.
15:54:19.90 [ERROR] 1 Exception encountered:

Engine traceback:
  in select
    ..
  in pants.core.goals.publish.run_publish
    `publish` goal
  in pants.core.goals.publish.package_for_publish
    ..
  in pants.core.goals.package.environment_aware_package
    ..
  in pants.backend.docker.goals.package_image.build_docker_image
    ..

Traceback (most recent call last):
  File "/home/runner/.cache/pants/setup/bootstrap-Linux-x86_64/2.15.1rc2_py38/lib/python3.8/site-packages/pants/engine/internals/selectors.py", line 593, in native_engine_generator_send
    res = func.send(arg)
  File "/home/runner/.cache/pants/setup/bootstrap-Linux-x86_64/2.15.1rc2_py38/lib/python3.8/site-packages/pants/backend/docker/goals/package_image.py", line 309, in build_docker_image
    raise ProcessExecutionFailure(
pants.engine.process.ProcessExecutionFailure: Process 'Building docker image ***.dkr.ecr.us-east-1.amazonaws.com/moz-helper-services:latest +1 additional tag.' failed with exit code 1.
stdout:

stderr:
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

</details>

As far as I can tell, enabling the `environments` feature confuses `pants` about where the `DOCKER_HOST` should be (our GitHub runners have `DOCKER_HOST` set to `unix:///run/docker/docker.sock` instead of the default `unix:///var/run/docker.sock`). We pass this environment variable down in our `pants.toml` file though:

<details>
  <summary>pants.toml</summary>

```toml
[docker]
env_vars = [
  "DOCKER_CONFIG=%(homedir)s/.docker",
  "DOCKER_DEFAULT_PLATFORM=linux/amd64",
  "HOME",
  "USER",
  "PATH",
  # used by action-runner-controller dind
  "DOCKER_CERT_PATH",
  "DOCKER_HOST",
  "DOCKER_TLS_VERIFY",
]
tools = [
  "dirname",
  "readlink",
  "python3",
  # These may be necessary if using Pyenv-installed Python.
  "cut",
  "sed",
  "bash",
  "sh",
]
default_repository = "{directory}"
build_args = ["PANTS_DISTRIBUTION_VERSION"]
```

</details>

For solutions to this we've tried upgrading to `pants==2.15.1rc2`, removing the `local_environment` completely and just using the `docker_environment`, and also using `pants.ci.toml` to [override which environment to use](https://www.pantsbuild.org/docs/environments#toggle-use-of-an-environment-for-some-consumers). All of this works within CI/CD when we specify no environments at all, but doesn't work locally. However it works locally when we *do* specify an environment, but doesn't work in CI/CD. 

**Pants version**
`2.15.0` / `2.15.1rc2`

**OS**
Linux (CI/CD)

**Additional info**
Add any other information about the problem here, such as attachments or links to gists, if relevant.

--- a/docs/notes/2.29.x.md
+++ b/docs/notes/2.29.x.md
@@ -18,6 +18,10 @@ Thank you to [Klaviyo](https://www.klaviyo.com/) for their Platinum tier support
 
 Adds a label to `docker_environment` containers so pantsd can cleanup any old, dead containers on startup. This prevents [old containers existing forever](https://github.com/pantsbuild/pants/issues/18307) in the event of a pantsd crash.
 
+Fixed an issue where environment targets with empty sequence fields would override global configuration instead of inheriting from it. This affected multiple backends including Docker, Python, and NodeJS. For the Docker backend, empty `docker_env_vars` fields in `docker_environment` targets would prevent inheritance from global `[docker].env_vars` settings, causing Docker buildx to fail due to missing required environment variables like `HOME`. (See [#20605](https://github.com/pantsbuild/pants/issues/20605))
+
+
+
 ### Goals
 
 ### Backends

--- a/src/python/pants/core/environments/rules.py
+++ b/src/python/pants/core/environments/rules.py
@@ -49,6 +49,7 @@ from pants.engine.intrinsics import docker_resolve_image
 from pants.engine.platform import Platform
 from pants.engine.rules import Get, QueryRule, collect_rules, concurrently, implicitly, rule
 from pants.engine.target import (
+    NO_VALUE,
     Field,
     FieldDefaultFactoryRequest,
     FieldDefaultFactoryResult,
@@ -765,20 +766,39 @@ def _add_option_field_for(
                 "value type."
             )
 
-    # The below class will never be used for static type checking outside of this function.
-    # so it's reasonably safe to use `ignore[name-defined]`. Ensure that all this remains valid
-    # if `_SIMPLE_OPTIONS` or `_LIST_OPTIONS` are ever modified.
-    class OptionField(field_type, _EnvironmentSensitiveOptionFieldMixin):  # type: ignore[valid-type, misc]
-        alias = f"{scope}_{snake_name}".replace("-", "_")
-        required = False
-        value: Any
-        help = (
+    class_attrs = {
+        "alias": f"{scope}_{snake_name}".replace("-", "_"),
+        "required": False,
+        "help": (
             f"Overrides the default value from the option `[{scope}].{snake_name}` when this "
             "environment target is active."
-        )
-        subsystem = env_aware_t
-        option_name = option.args[0]
+        ),
+        "subsystem": env_aware_t,
+        "option_name": option.args[0],
+    }
 
+    # For sequence fields, use special handling to distinguish "not specified" from "explicitly empty"
+    if issubclass(field_type, SequenceField):
+        class_attrs["none_is_valid_value"] = True
+
+        def compute_value(cls, raw_value: Any | None, address: Address) -> Any | None:
+            # When field is not specified in BUILD file, return None to indicate inheritance
+            if raw_value is NO_VALUE:
+                return None
+            # When explicitly set to empty list, return empty tuple (don't inherit)
+            elif raw_value == []:
+                return ()
+            # For other values, use normal field processing
+            else:
+                return field_type.compute_value(raw_value, address)
+
+        class_attrs["compute_value"] = classmethod(compute_value)
+
+    # Create the OptionField class dynamically
+    OptionField = cast(
+        "type[Field]",
+        type("OptionField", (field_type, _EnvironmentSensitiveOptionFieldMixin), class_attrs),
+    )
     setattr(OptionField, "__qualname__", f"{option_type.__qualname__}.{OptionField.__name__}")
 
     return [
@@ -804,13 +824,6 @@ def resolve_environment_sensitive_option(name: str, subsystem: Subsystem.Environ
 
     maybe = options.get((type(subsystem), name))
     if maybe is None or maybe.value is None:
-        return None
-
-    # For sequence fields, treat empty sequences as "not specified" to allow fallback
-    # to global configuration. This handles cases where environment fields are created
-    # with default empty values (e.g., () for StringSequenceField) but user didn't
-    # explicitly set any values.
-    if isinstance(maybe, SequenceField) and not maybe.value:
         return None
 
     return maybe.value

--- a/src/python/pants/core/environments/rules.py
+++ b/src/python/pants/core/environments/rules.py
@@ -53,6 +53,7 @@ from pants.engine.target import (
     FieldDefaultFactoryRequest,
     FieldDefaultFactoryResult,
     FieldSet,
+    SequenceField,
     StringField,
     StringSequenceField,
     Target,
@@ -804,8 +805,15 @@ def resolve_environment_sensitive_option(name: str, subsystem: Subsystem.Environ
     maybe = options.get((type(subsystem), name))
     if maybe is None or maybe.value is None:
         return None
-    else:
-        return maybe.value
+
+    # For sequence fields, treat empty sequences as "not specified" to allow fallback
+    # to global configuration. This handles cases where environment fields are created
+    # with default empty values (e.g., () for StringSequenceField) but user didn't
+    # explicitly set any values.
+    if isinstance(maybe, SequenceField) and not maybe.value:
+        return None
+
+    return maybe.value
 
 
 @memoized

--- a/src/python/pants/core/environments/rules.py
+++ b/src/python/pants/core/environments/rules.py
@@ -785,12 +785,8 @@ def _add_option_field_for(
             # When field is not specified in BUILD file, return None to indicate inheritance
             if raw_value is NO_VALUE:
                 return None
-            # When explicitly set to empty list, return empty tuple (don't inherit)
-            elif raw_value == []:
-                return ()
             # For other values, use normal field processing
-            else:
-                return field_type.compute_value(raw_value, address)
+            return field_type.compute_value(raw_value, address)
 
         class_attrs["compute_value"] = classmethod(compute_value)
 

--- a/src/python/pants/core/environments/rules_test.py
+++ b/src/python/pants/core/environments/rules_test.py
@@ -6,6 +6,7 @@ from __future__ import annotations
 from collections.abc import Sequence
 from dataclasses import dataclass
 from textwrap import dedent
+from typing import Any
 
 import pytest
 
@@ -22,6 +23,7 @@ from pants.core.environments.rules import (
     UnrecognizedEnvironmentError,
     extract_process_config_from_environment,
     resolve_environment_name,
+    resolve_environment_sensitive_option,
 )
 from pants.core.environments.subsystems import EnvironmentsSubsystem
 from pants.core.environments.target_types import (
@@ -44,6 +46,8 @@ from pants.engine.platform import Platform
 from pants.engine.process import ProcessCacheScope
 from pants.engine.target import FieldSet, OptionalSingleSourceField, Target
 from pants.option.global_options import GlobalOptions, KeepSandboxes
+from pants.option.option_types import ShellStrListOption
+from pants.option.subsystem import Subsystem
 from pants.testutil.option_util import create_subsystem
 from pants.testutil.rule_runner import (
     MockGet,
@@ -555,3 +559,64 @@ def test_find_chosen_local_and_experimental_workspace_environments(rule_runner: 
 
     chosen_workspace_env = rule_runner.request(ChosenLocalWorkspaceEnvironmentName, [])
     assert chosen_workspace_env.val.val == "workspace"
+
+
+def test_empty_sequence_field_override_bug() -> None:
+    """Test that empty sequence fields don't override global configuration."""
+
+    class TestSubsystem(Subsystem):
+        options_scope = "test-subsystem"
+        help = "Test subsystem"
+
+        class EnvironmentAware(Subsystem.EnvironmentAware):
+            test_sequence_option = ShellStrListOption(help="Test sequence option")
+
+    rule_runner = RuleRunner(
+        target_types=[LocalEnvironmentTarget],
+        rules=[
+            *TestSubsystem.rules(),
+            *environments_rules.rules(),
+        ],
+    )
+
+    rule_runner.set_options(
+        [
+            "--test-subsystem-test-sequence-option=GLOBAL_VAR=global_value",
+            "--test-subsystem-test-sequence-option=TEST_VAR=test_value",
+        ]
+    )
+
+    rule_runner.write_files(
+        {
+            "BUILD": dedent("""\
+            local_environment(
+                name="test_env",
+                compatible_platforms=["linux_x86_64", "macos_arm64"],
+            )
+            """),
+        }
+    )
+
+    env_target = rule_runner.get_target(Address("", target_name="test_env"))
+
+    test_sequence_field: Any = None
+    for field_type in env_target.field_values:
+        field = env_target.field_values[field_type]
+        if hasattr(field, "alias") and field.alias == "test_subsystem_test_sequence_option":
+            test_sequence_field = field
+            break
+
+    assert test_sequence_field is not None, "Dynamic field creation not working"
+
+    assert test_sequence_field.value == ()
+
+    subsystem_class = test_sequence_field.subsystem
+    mock_env_aware = subsystem_class()
+    mock_env_aware.env_tgt = EnvironmentTarget("test_env", env_target)
+
+    result = resolve_environment_sensitive_option(
+        test_sequence_field.option_name,
+        mock_env_aware,
+    )
+
+    assert result is None, f"Expected None for empty sequence field, got {result}"

--- a/src/python/pants/core/environments/rules_test.py
+++ b/src/python/pants/core/environments/rules_test.py
@@ -561,8 +561,8 @@ def test_find_chosen_local_and_experimental_workspace_environments(rule_runner: 
     assert chosen_workspace_env.val.val == "workspace"
 
 
-def test_empty_sequence_field_override_bug() -> None:
-    """Test that empty sequence fields don't override global configuration."""
+def test_sequence_field_inheritance() -> None:
+    """Test environment field inheritance for sequence fields."""
 
     class TestSubsystem(Subsystem):
         options_scope = "test-subsystem"
@@ -586,37 +586,82 @@ def test_empty_sequence_field_override_bug() -> None:
         ]
     )
 
-    rule_runner.write_files(
-        {
-            "BUILD": dedent("""\
+    def test_environment_resolution(
+        runner: RuleRunner,
+        env_name: str,
+        env_definition: str,
+        expected_result: Any,
+        description: str,
+    ):
+        runner.write_files(
+            {
+                f"{env_name}/BUILD": dedent(f"""\
             local_environment(
-                name="test_env",
+                name="{env_name}",
                 compatible_platforms=["linux_x86_64", "macos_arm64"],
+                {env_definition}
             )
-            """),
-        }
+            """)
+            }
+        )
+
+        env_target = runner.get_target(Address(env_name, target_name=env_name))
+
+        test_sequence_field: Any = None
+        for field_type in env_target.field_values:
+            field = env_target.field_values[field_type]
+            if hasattr(field, "alias") and field.alias == "test_subsystem_test_sequence_option":
+                test_sequence_field = field
+                break
+
+        assert test_sequence_field is not None
+
+        subsystem_class = test_sequence_field.subsystem
+        mock_env_aware = subsystem_class()
+        mock_env_aware.env_tgt = EnvironmentTarget(env_name, env_target)
+
+        result = resolve_environment_sensitive_option(
+            test_sequence_field.option_name,
+            mock_env_aware,
+        )
+
+        assert result == expected_result, f"{description}: Expected {expected_result}, got {result}"
+
+    test_environment_resolution(
+        rule_runner, "env1", "", None, "Field not specified - should inherit global config"
     )
 
-    env_target = rule_runner.get_target(Address("", target_name="test_env"))
-
-    test_sequence_field: Any = None
-    for field_type in env_target.field_values:
-        field = env_target.field_values[field_type]
-        if hasattr(field, "alias") and field.alias == "test_subsystem_test_sequence_option":
-            test_sequence_field = field
-            break
-
-    assert test_sequence_field is not None, "Dynamic field creation not working"
-
-    assert test_sequence_field.value == ()
-
-    subsystem_class = test_sequence_field.subsystem
-    mock_env_aware = subsystem_class()
-    mock_env_aware.env_tgt = EnvironmentTarget("test_env", env_target)
-
-    result = resolve_environment_sensitive_option(
-        test_sequence_field.option_name,
-        mock_env_aware,
+    test_environment_resolution(
+        rule_runner,
+        "env2",
+        "test_subsystem_test_sequence_option=[],",
+        (),
+        "Global values can be overridden with empty",
     )
 
-    assert result is None, f"Expected None for empty sequence field, got {result}"
+    test_environment_resolution(
+        rule_runner,
+        "env3",
+        'test_subsystem_test_sequence_option=["ENV_VAR=env_value"],',
+        ("ENV_VAR=env_value",),
+        "Field with values - should override global config",
+    )
+
+    # Set global options to empty to test inheritance from empty global
+    rule_runner.set_options([])
+
+    test_environment_resolution(
+        rule_runner,
+        "env4",
+        "",
+        None,
+        "Field not specified with empty global - should inherit empty global",
+    )
+
+    test_environment_resolution(
+        rule_runner,
+        "env5",
+        'test_subsystem_test_sequence_option=["NEW_VAR=new_value"],',
+        ("NEW_VAR=new_value",),
+        "Empty global can be overridden with values",
+    )

